### PR TITLE
refactor to support msg inside msg, constant, static array initialize

### PR DESCRIPTION
**Pull request by yuokamoto**
- use msg inside msg instead of expanding all msgs
- Add namespace as prefix for msg name. Prefix no need to be same as original namespace and defined in yaml file.
- Add word conversion. Support conveting word such as Covariance -> Cov. This is used to avoid too long class/struct name to avoid breaking UE maximum class name. https://forums.unrealengine.com/t/qeustion-about-class-name-length-limitation-when-adding-code-through-the-editor/49552 
- Support initializing fix sized array
- Define constant as constexpr and CONST_<const name> function to used in BP.
- use functions in rclUtilities in rclUE

related pr:
https://github.com/rapyuta-robotics/rclUE/pull/59

### CODE REVIEW

File: BuildROS2/setup_ros2_pkgs.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@ ROS2_WS=$(pwd)/ros2_ws
 PKGS=$2
 
 # install dependency and get ros2 sources
-./pre_build_ros2_pkgs.sh
+# ./pre_build_ros2_pkgs.sh
```
Comment (Tadinu) [inline BuildROS2/setup_ros2_pkgs.sh]: no need this anymore?

File: CodeGen/gen_ue_from_ros.py [hunk 2]
```diff
@@ -24,51 +25,102 @@
...
     'byte': 'uint8',
     'char': 'uint8',
     'float32': 'float',
     'float64': 'double',
     'string': 'FString',
-    'Vector3':'FVector',
-    'Point': 'FVector',
-    'Point32':'FVector', 
-    'Quaternion': 'FQuat'
+    # 'wstring': 'std::u16string', #donot works
```
Comment (Tadinu) [inline CodeGen/gen_ue_from_ros.py]: FString supports both normal and wide string, so we could use it here I suppose
https://www.coconutlizard.co.uk/blog/strings-and-other-things

File: CodeGen/gen_ue_from_ros.py [hunk 3]
```diff
@@ -79,30 +131,56 @@
...
+        t = f'TArray<{convert_to_ue_type(t, pkgs_name_mapping, name_mapping)[0]}>'
     elif ('[' in t) and (']' in t):
         tmp = re.split('\[|\]', t)
         size = tmp[1].replace('<=','')
-        t = f'TArray<{convert_to_ue_type(tmp[0])[0]}>'
+        t = f'TArray<{convert_to_ue_type(tmp[0], pkgs_name_mapping, name_mapping)[0]}>'
+    elif t in TYPE_CONVERSION :
+        if t == 'wstring':
+            logger.error('wstring is not supported. this will cause compilation error in UE.' + str(original_names) + ' (get_ue_var_name)')
+        t = TYPE_CONVERSION[t]
```
Comment (Tadinu) [inline CodeGen/gen_ue_from_ros.py]: To be converted to FString if possible?

File: CodeGen/templates/Action.h [hunk 4]
```diff
@@ -107,43 +120,53 @@ struct {{data.ModuleAPI}} F{{data.StructName}}GetResultRequest
...
 	UPROPERTY()
-	int8 Status = 0;
+	int8 GRResStatus = 0;
 
 	{{data.ResultTypes}}
 
+	FROS{{data.UEName}}GRRes()
+    {
+		{{data.ResultConstructor}}
+    }
```
Comment (Tadinu) [inline CodeGen/templates/Action.h:138]: Could it be in initialization list like `FROS{{data.UEName}}GRRes():{{data.ResultConstructor}}`?

File: CodeGen/templates/Action.h [hunk 5]
```diff
@@ -3,30 +3,43 @@
...
+  	
+	TArray<uint, TFixedAllocator<16>> GoalId;
 
 	{{data.GoalTypes}}
 
-	F{{data.StructName}}SendGoalRequest()
+	FROS{{data.UEName}}SGReq()
     {
         UROS2Utils::GenerateRandomUUID16(GoalId);
+		{{data.GoalConstructor}}
```
Comment (Tadinu) [inline CodeGen/templates/Action.h:42]: Could it be in initialization list like `FROS{{data.UEName}}SGReq():{{data.GoalConstructor}}?`

File: CodeGen/templates/Action.h [hunk 6]
```diff
@@ -107,43 +120,53 @@ struct {{data.ModuleAPI}} F{{data.StructName}}GetResultRequest
...
   	TArray<uint, TFixedAllocator<16>> GoalId;
 
 	{{data.FeedbackTypes}}
 
 
-    F{{data.StructName}}FeedbackMessage()
+    FROS{{data.UEName}}FB()
     {
         GoalId.Init(0, 16);
+		{{data.FeedbackConstructor}}
```
Comment (Tadinu) [inline CodeGen/templates/Action.h:169]: same as above as for initialization list?

File: CodeGen/templates/Msg.h [hunk 7]
```diff
@@ -3,23 +3,37 @@
 
 #pragma once
 
+// UE
 #include <CoreMinimal.h>
```
Comment (Tadinu) [inline CodeGen/templates/Msg.h]:
```suggestion
#include "CoreMinimal.h"
```
Just to make it consistent with all UE header inclusion

File: CodeGen/templates/Srv.h [hunk 8]
```diff
@@ -3,23 +3,39 @@
 
 #pragma once
 
+// UE
 #include <CoreMinimal.h>
```
Comment (Tadinu) [inline CodeGen/templates/Srv.h]:
```suggestion
#include "CoreMinimal.h"
```

**Review by Tadinu — APPROVED** [review]

File: CodeGen/templates/Action.h (hunk 5 above)
Comment (yuokamoto) [reply CodeGen/templates/Action.h:42]: it is not constructor call but initialization step such as. I added this to initialize fixed size array
image

Comment (Tadinu) [reply CodeGen/templates/Action.h:42]: I see.., btw, should it be `Covariance.SetNumZeroed(36);` only, without for loop?

Comment (yuokamoto) [reply CodeGen/templates/Action.h:42]: ah, true. will fix

File: build_install_codegen.py [hunk 9]
```diff
@@ -14,7 +15,130 @@ def managed_chdir(path):
         yield 
     finally:
         os.chdir(prev_path)
+
+def load_from_conifgs(file_names):
```
Comment (tetorea) [inline build_install_codegen.py]:
```suggestion
def load_from_configs(file_names):
```

File: build_install_codegen.py [hunk 10]
```diff
@@ -41,30 +165,19 @@ def managed_chdir(path):
...
+        config_files.extend(args.config)
 
-    home = os.environ['HOME']
-    projectPath = os.path.join(home, config['args']['ue_proj_path'])
-    pluginName = config['args']['ue_plugin_name']
-    pluginFolderName = config['args']['ue_plugin_folder_name']
-    ros_pkgs = config['target_pkgs']
+    UEPath, projectPath, pluginName, \
+        pluginFolderName, targetThirdpartyFolderName, \
+        target, black_list, dependency, name_mapping = load_from_conifgs(config_files)
```
Comment (tetorea) [inline build_install_codegen.py]:
```suggestion
        target, black_list, dependency, name_mapping = load_from_configs(config_files)
```

**Review by tetorea — APPROVED** [review]

**Pull request merged**